lib/Dialect/FIRRTL/Ops.cpp

```diff
@@ -74,6 +74,21 @@ static LogicalResult verifyCircuitOp(CircuitOp &circuit) {
     return failure();
   }
 
+  // Check that no extmodule defname conflicts with another module
+  // name in the circuit.
+  for (auto &op : *circuit.getBody()) {
+    if (auto extModule = dyn_cast<FExtModuleOp>(op)) {
+      auto defname = extModule.defnameAttr();
+      if (defname && circuit.lookupSymbol(defname.getValue()) &&
+          isa<FModuleOp>(circuit.lookupSymbol(defname.getValue()))) {
```

Comment:
Plz pull the `circuit.lookupSymbol(defname.getValue())` out into its own expression -- no need to do the lookup twice.

Is there also a potential problem with defnames of two different ext modules conflicting with each other?

If so, you might create a `DenseMap<Attr,Op>`, and keep throwing the defnames into the set for each of these things as you go.  when a conflict is detected, you can emit a 'note' pointing to the declaration that it conflicts with.

Reply:
Sure, I can rework this. Nice suggestion.

The `ExtModule` conflict resolution logic is more complex and I was planning to do that in a separate PR. 

Roughly two `ExtModule`s that have the same defname and _do not_ have parameters, must have the same ports, in the same order, with the same types, and the same widths. Two `ExtModule`s that have the same defname, but take parameters, must have the same ports, in the same order, and same types, but widths may be different. (The latter can have false negatives, but errs on the side of safety because the parameters may affect port widths.)

Some code examples of tests in the Scala FIRRTL compiler are here: https://github.com/freechipsproject/firrtl/pull/1734/files#diff-a34ade886920ac99fc85ef192d5fbe3397d3a8364d5799ecc22913388e16af30. My plan was to add all of these to CIRCT FIRRTL Dialect testing.

Reply:
cool sounds good!

Reply:
Removed the repeated symbol lookup in a later commit. I'll opt for the `DenseMap` approach in a follow-up PR to check the more subtle `defname` collision possibilities between `ExtModule`s. Thanks for the pointer to `DenseMap`. (I know my way around the Scala standard library, but what LLVM provides is a learning process.)